clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -23,178 +23,419 @@ package common
-		objectIndex := objectCounter
-		wg.Start(func() {
-			if err := waitForRuntimeObject(config.ClientSet, runtimeObjectsList[objectIndex], timeout); err != nil {
-				errList.Append(fmt.Errorf("waiting for %v error: %v", kind, err))
+	w.isRunning = true
+	w.stopCh = make(chan struct{})
+	w.informer = cache.NewSharedInformer(listerWatcher, obj, 0)
+	w.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			runtimeObj, ok := obj.(runtime.Object)
+			if !ok {
 				return
 			}
-
-			atomic.AddInt32(&runningRuntimeObjects, 1)
-			objName, err := getNameFromRuntimeObject(runtimeObjectsList[objectIndex])
+			resourceVersion, err := getResourceVersionFromRuntimeObject(runtimeObj)
 			if err != nil {
-				errList.Append(fmt.Errorf("reading object name error: %v", err))
 				return
 			}
-			objNamespace, err := getNamespaceFromRuntimeObject(runtimeObjectsList[objectIndex])
+			w.updateOpResourceVersion(resourceVersion)
+			w.handleObject(c, runtimeObj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldRuntimeObj, ok := oldObj.(runtime.Object)
+			if !ok {
+				return
+			}
+			newRuntimeObj, ok := newObj.(runtime.Object)
+			if !ok {
+				return
+			}
+			resourceVersion, err := getResourceVersionFromRuntimeObject(newRuntimeObj)
 			if err != nil {
-				errList.Append(fmt.Errorf("reading object namespace error: %v", err))
 				return
 			}
-			glog.Infof("%s: %s (%s) has all pods running", waitForControlledPodsRunningName, objName, objNamespace)
-		})
+			// Increment version regardless of change in spec.
+			w.updateOpResourceVersion(resourceVersion)
+			oldSpec, err := getSpecFromRuntimeObject(oldRuntimeObj)
+			if err != nil {
+				return
+			}
+			newSpec, err := getSpecFromRuntimeObject(newRuntimeObj)
+			if err != nil {
+				return
+			}
+			if reflect.DeepEqual(oldSpec, newSpec) {
+				// If spec of the object hasn't change, do nothing.
+				return
+			}
+			w.handleObject(c, newRuntimeObj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			runtimeObj, ok := obj.(runtime.Object)
+			if !ok {
+				return
+			}
+			resourceVersion, err := getResourceVersionFromRuntimeObject(runtimeObj)
+			if err != nil {
+				return
+			}
+			w.updateOpResourceVersion(resourceVersion)
+			w.deleteObject(runtimeObj)
+		},
+	})
+
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
 	}
-	wg.Wait()
+	timeoutTimer.Stop()
 
-	if int(runningRuntimeObjects) == len(runtimeObjectsList) {
-		glog.Infof("%s: %d / %d %ss are running with all pods", waitForControlledPodsRunningName, runningRuntimeObjects, len(runtimeObjectsList), kind)
-		return summaries, nil
+	store := w.informer.GetStore()
+	objects := store.List()
+	for i := range objects {
```

Comment:
To we need it?
Won't informer send all Add events at the beginning anyway?

Reply:
TBH, I want to keep the sync part.
I agree that in our case it may be not needed. However, I want this function to be more generic, able to handle many cases. IMHO if somebody uses this measurement when RCs already exist it should behave correctly.

Reply:
You are right - I removed this check.
I ran a test and it behaved as you decried. However, I haven't found the code responsible for this yet.